### PR TITLE
WIP: Use GlyphData.xml

### PR DESCRIPTION
**Pull request by madig**
Instead of making https://github.com/googlei18n/glyphsLib/pull/403 work, include GlyphData.xml and just parse it.

### CODE REVIEW

**Comment by madig** [comment]
Hm, could be something to do after we figure out the basics´.

**Comment by madig** [comment]
Note to self: make data dicts return empty dict so i don't have to do `or {}` everywhere.

**Comment by anthrotype** [comment]
> make data dicts return empty dict so i don't have to do or {} everywhere

just do `data.get(key, {})`

**Comment by madig** [comment]
I did some benchmarking with the following script (some things commented out because Unicode behavior is different now):
```Python
from glyphsLib.glyphdata import get_glyph

prod = lambda n: get_glyph(n).production_name
cat = lambda n: (get_glyph(n).category, get_glyph(n).subCategory)
uni = lambda n: get_glyph(n).unicode


def test():
    assert prod(".notdef") == ".notdef"
    assert prod("eacute") == "eacute"
    assert prod("Abreveacute") == "uni1EAE"
    assert prod("C-fraktur") == "uni212D"
    assert prod("Dboldscript-math") == "u1D4D3"
    assert prod("fi") == "fi"
    assert prod("s_t") == "s_t"
    assert prod("Gcommaaccent") == "uni0122"
    assert prod("o_f_f_i.foo") == "o_f_f_i.foo"
    assert prod("ain_alefMaksura-ar.fina") == "uniFD13"
    assert prod("brevecomb") == "uni0306"
    assert prod("brevecomb.case") == "uni0306.case"
    assert prod("brevecomb_acutecomb") == "uni03060301"
    assert prod("brevecomb_acutecomb.case") == "uni03060301.case"
    assert prod("brevecomb_a_a_a") == "uni0306006100610061"
    assert prod("brevecomb_a_a_a.case") == "uni0306006100610061.case"
    assert prod("brevecomb_aaa.case") == "brevecomb_aaa.case"
    assert prod("brevecomb_Dboldscript-math") == "uni0306_u1D4D3"
    assert prod("brevecomb_Dboldscript-math.f.r") == "uni0306_u1D4D3.f.r"
    assert prod("Dboldscript-math_Dboldscript-math") == "u1D4D3_u1D4D3"
    assert prod("Dboldscript-math_Dboldscript-math.f") == "u1D4D3_u1D4D3.f"
    assert prod("Dboldscript-math_a") == "u1D4D3_a"
    assert prod("a_Dboldscript-math") == "a_u1D4D3"
    assert prod("Dboldscript-math_a_aa") == "u1D4D3_a_uniA733"
    # assert prod("Dboldscript-math_a_aaa") == "Dboldscriptmath_a_aaa"
    assert prod("brevecomb_Dboldscript-math") == "uni0306_u1D4D3"
    assert prod("Dboldscript-math_brevecomb") == "u1D4D3_uni0306"
    assert prod("idotaccent") == "i.loclTRK"
    assert prod("a_idotaccent") == "a_i.loclTRK"
    assert prod("a_idotaccent_a") == "a_idotaccent_a"
    assert prod("a_a_acutecomb") == "a_a_acutecomb"
    assert prod("a_a_dieresiscomb") == "uni006100610308"
    assert prod("brevecomb_acutecomb") == "uni03060301"
    assert prod("vaphalaa-malayalam") == "uni0D030D35.1"

    # assert uni(".notdef") is None
    # assert uni("eacute") == "00E9"
    # assert uni("Abreveacute") == "1EAE"
    # assert uni("C-fraktur") == "212D"
    # assert uni("Dboldscript-math") == "1D4D3"
    # assert uni("fi") == "FB01"
    # assert uni("Gcommaaccent") == "0122"
    # assert uni("s_t") is None
    # assert uni("o_f_f_i.foo") is None
    # assert uni("brevecomb") == "0306"
    # assert uni("brevecomb.case") is None
    # assert uni("brevecomb_acutecomb") is None
    # assert uni("brevecomb_acutecomb.case") is None

    assert cat(".notdef") == ("Separator", None)
    assert cat("uni000D") == ("Separator", None)
    assert cat("boxHeavyUp") == ("Symbol", "Geometry")
    assert cat("eacute") == ("Letter", "Lowercase")
    assert cat("Abreveacute") == ("Letter", "Uppercase")
    assert cat("C-fraktur") == ("Letter", "Uppercase")
    assert cat("fi") == ("Letter", "Ligature")
    assert cat("fi.alt") == ("Letter", "Ligature")
    assert cat("hib-ko") == ("Letter", "Syllable")
    assert cat("one.foo") == ("Number", "Decimal Digit")
    assert cat("one_two.foo") == ("Number", "Ligature")
    assert cat("o_f_f_i") == ("Letter", "Ligature")
    assert cat("o_f_f_i.foo") == ("Letter", "Ligature")
    assert cat("ain_alefMaksura-ar.fina") == ("Letter", "Ligature")
    assert cat("brevecomb") == ("Mark", "Nonspacing")
    assert cat("brevecomb.case") == ("Mark", "Nonspacing")
    assert cat("brevecomb_acutecomb") == ("Mark", "Nonspacing")
    assert cat("brevecomb_acutecomb.case") == ("Mark", "Nonspacing")

    u, g = get_glyph("uni07F0"), get_glyph("longlowtonecomb-nko")
    assert (u.category, g.category) == ("Mark", "Mark")
    assert (u.subCategory, g.subCategory) == ("Nonspacing", "Nonspacing")
    assert (u.production_name, g.production_name) == ("uni07F0", "uni07F0")
    #assert (u.unicode, g.unicode) == ("07F0", "07F0")


test()
```
I uninstalled the global, baseline glyphsLib 3.0.3 and deleted this branch's venv, then rebooted, and reinstalled the baseline glyphsLib and created a fresh venv. In IPython, I ran `%timeit %run -i "test.py"`, one time outside and one time inside the venv. Results:

Results:
```
Baseline:
In [1]: %timeit %run -i "..\tmp\tet.py"
1.52 ms ± 130 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [2]: %timeit %run -i "..\tmp\tet.py"
1.59 ms ± 19.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

This branch:
In [1]: %timeit %run -i "..\tmp\tet.py"
1.49 ms ± 14.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [2]: %timeit %run -i "..\tmp\tet.py"
1.48 ms ± 14 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

Baseline:
In [1]: %timeit %run -i "..\tmp\tet.py"
1.59 ms ± 16.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [2]: %timeit %run -i "..\tmp\tet.py"
1.59 ms ± 18.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

This branch:
In [1]: %timeit %run -i "..\tmp\tet.py"
1.48 ms ± 29.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
So this branch seems to be faster? I'm not seeing cold-start slowness for this branch so I'm not sure how scientifically rigorous this is.

**Comment by madig** [comment]
I installed this branch in a fresh venv (no `-e`) and got
```
In [2]: %timeit %run -i "C:\\UsersLocal\\nikolaus.waxweiler\\Development\\tmp\\tet.py"
1.51 ms ± 12.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [3]: %timeit %run -i "C:\\UsersLocal\\nikolaus.waxweiler\\Development\\tmp\\tet.py"
1.5 ms ± 11.5 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

**Comment by madig** [comment]
Including it yields for this branch:
```
In [1]: %timeit %run -i "C:\\UsersLocal\\nikolaus.waxweiler\\Development\\tmp\\tet.py"
1.58 ms ± 216 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [2]: %timeit %run -i "C:\\UsersLocal\\nikolaus.waxweiler\\Development\\tmp\\tet.py"
1.52 ms ± 13 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
So, slightly faster than baseline I suppose.

File: Lib/glyphsLib/glyphdata.py [hunk 1]
```diff
@@ -14,83 +14,181 @@
...
+                os.path.dirname(glyphsLib.__file__),
+                "data",
+                "GlyphData_Ideographs.xml",
+            ),
+        )
 
-    # First, get the base name of the glyph. .notdef and .null are exceptions.
-    # Periods denote glyph variants as per the AGLFN convention, which should
-    # be in the same category as their base glyph.
-    if glyph_name in (".notdef", ".null"):
```
Comment (anthrotype) [inline Lib/glyphsLib/glyphdata.py]: do we not need to special-case '.notdef' and '.null' any more here?

Comment (madig) [reply Lib/glyphsLib/glyphdata.py]: Both are part of the XML data.

**Comment by anthrotype** [comment]
this PR would fix https://github.com/googlei18n/glyphsLib/issues/141

**Comment by anthrotype** [comment]
and would fix this as well https://github.com/googlei18n/glyphsLib/issues/402

shall we also remove that generate_glyphdata.py script?

**Comment by madig** [comment]
Ah yes, will remove it.

**Pull request merged**